Move edit role appointments page to GDS Design System 

chriswhitehouse-tw opened the pull request:
## Description

This ports the role appointments edit page to the GOV.UK Design System. Then adds the new view in the GOV.UK Design System and conditionally renders the Bootstrap or GOV.UK Design System based on whether the user has the "Preview design system" permission.

## Screenshots

### Before

#### On landing (can delete, no speeches)
![Screenshot 2023-03-31 at 15 39 33](https://user-images.githubusercontent.com/91492293/229151841-f713a735-cd69-4b5b-8e77-1a7e9e5773d0.png)

#### On landing (can't delete, with speeches)
![Screenshot 2023-03-31 at 15 37 47](https://user-images.githubusercontent.com/91492293/229152032-ad0d03de-a5f3-46e9-9217-1b2c62ba14d0.png)

#### Error (no started at date, no ended at date)
![Screenshot 2023-03-31 at 15 38 19](https://user-images.githubusercontent.com/91492293/229152161-e2651e9c-d5ab-4a7b-91c5-eb0d8d6a8f53.png)

#### Error (no started at date, but with end date)
![Screenshot 2023-03-31 at 15 38 37](https://user-images.githubusercontent.com/91492293/229152326-2c7863ee-199b-4ee4-a87c-77c453c9237c.png)

### After

#### On landing (can delete, no speeches)
<img width="1792" alt="Screenshot 2023-04-05 at 17 00 39" src="https://user-images.githubusercontent.com/91492293/230138438-86a3fb23-179c-4339-a999-5cd80f2e4d1f.png">

#### On landing (can't delete, with speeches)
![Screenshot 2023-04-06 at 11 28 46](https://user-images.githubusercontent.com/91492293/230351756-81df6ac8-e49e-4b9f-9b91-1d83ac7b78f8.png)

#### On landing (single option on drop down for edit)
![Screenshot 2023-04-06 at 11 31 02](https://user-images.githubusercontent.com/91492293/230351843-82c859cd-8500-4eeb-b83f-c13af8a2aabb.png)

#### Error (no started at date, no ended at date)
<img width="1792" alt="Screenshot 2023-04-05 at 17 02 03" src="https://user-images.githubusercontent.com/91492293/230138643-5e0d9ecc-a478-4b7b-831c-fc24e8439ce3.png">

#### Error (no started at date, but with end date)
<img width="1792" alt="Screenshot 2023-04-05 at 17 02 24" src="https://user-images.githubusercontent.com/91492293/230138673-df6b8c85-7cfa-49c5-8bb3-372c41646a8a.png">

## Trello

https://trello.com/c/R5A32Kau

Add content_for page_title, title, and error_summary

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

chriswhitehouse-tw commented:
Couple of areas I could do with specific guidance:

* Is the single drop down option ok for the select? I was unsure how to disable select.
* Rendering of the 'none' for speeches, and 'Cannot delete' doesn't look great. These are both p tags. Is there an alternative  styling I should be using?
* There is a server error if a user tries to submit a blank started at date, but populated ended at date. Presumably to do with model validations, but not sure why the error summary is not handling this?

davidgisbey commented on app/views/admin/role_appointments/edit.html.erb:
```diff
@@ -1,40 +1,74 @@
+<% content_for :page_title, "Edit appointment to #{@role_appointment.role.name}"  %>
+<% content_for :title, "Edit appointment to #{@role_appointment.role.name}" %>
+<% content_for :title_margin_bottom, 6 %>
+<% content_for :error_summary, render(Admin::ErrorSummaryComponent.new(object: @role_appointment, parent_class: "role_appointment")) %>
```
```
<% content_for :error_summary, render(Admin::ErrorSummaryComponent.new(object: @role_appointment)) %>
```

This won't break anything, but we only need to pass a `parent_class` in when the model is inheriting from a parent class and the fields use the parent class.

For example, we use one on the edit edition page as the names of the fields are `edition[title]` etc. but the class of the objects if you call`@edition.class` is `Consultation, Case Study` etc. 

Without passing in the parent_class to the error summary it would look for fields names `consultation[title]` etc.

Let me know if that doesn't make sense

chriswhitehouse-tw replied:
Cool, yes that makes perfect sense. I can see how the ErrorSummaryComponent handles this now.

davidgisbey commented on app/views/admin/role_appointments/edit.html.erb:
```diff
@@ -1,40 +1,74 @@
-
-      <%= form.save_or_cancel cancel: edit_admin_role_path(@role_appointment.role) %>
+      <div class="govuk-button-group">
+        <%= render "govuk_publishing_components/components/button", {
```
I think we can make this a bit simpler

```
  <div class="govuk-button-group">
        <%= render "govuk_publishing_components/components/button", {
          text: "Save",
          data_attributes: {
            module: "gem-track-click",
            "track-category": "form-button",
            "track-action": "role-appointment-button",
            "track-label": "Save"
          }
        } %>
```

If you look a the [component](https://components.publishing.service.gov.uk/component-guide/button), all you need to provide is text and it defaults to having a type of submit etc.

Providing a name/value is only required when the controller is figuring out where to send a user submits a form when there are 2 save buttons. So it can do something like

```
if object.save 
  if params[:save] == "save and continue"
    redirect to x
  else
    redirect to x
  end
else 
  render same page with errors
end
```

chriswhitehouse-tw replied:
Cool, yes, that makes perfect sense as well. That was me being lazy and copy pasting over from the role edit page, without thinking hard about whether each of those parameters are actually required.

davidgisbey replied:
No worries happens to me all  the time

davidgisbey commented on app/views/admin/role_appointments/edit.html.erb:
```diff
@@ -1,40 +1,74 @@
 
     <% if @role_appointment.destroyable? %>
-      <%= button_to "Delete", admin_role_appointment_path(@role_appointment), method: :delete, class: "btn btn-danger" %>
+      <%= form_for [:admin, @role_appointment], method: "delete" do |form| %>
```
Sorry we should put this in the docs but we should add confirm delete pages for destroying objects. 

To follow the established pattern i 'd think this should have a red delete link inline between the save button and cancel link

chriswhitehouse-tw replied:
Thanks, yup, have updated to reflect this.

davidgisbey replied:
Thanks! Sorry we should add this to the card when required.

davidgisbey commented on app/views/admin/role_appointments/edit.html.erb:
```diff
@@ -1,40 +1,74 @@
+    } %>
     <% if @role_appointment.other_appointments_for_same_role.any? %>
-      <%= render partial: "admin/role_appointments/legacy_list", locals: { role_appointments: @role_appointment.other_appointments_for_same_role }%>
+      <%= render "admin/role_appointments/list", { role_appointments: @role_appointment.other_appointments_for_same_role }%>
```
<%= render "admin/role_appointments/list",  role_appointments: @role_appointment.other_appointments_for_same_role %>

chriswhitehouse-tw replied:
cool, yup, good spot, thanks.

davidgisbey commented on app/views/admin/role_appointments/edit.html.erb:
```diff
@@ -1,40 +1,74 @@
+            end
+        } %>
       <% else %>
         <p>None</p>
```
<p class="govuk-body>None</p>

chriswhitehouse-tw replied:
cool, thank you. That looks better.

davidgisbey commented on test/functional/admin/role_appointments_controller_test.rb:
```diff
@@ -2,7 +2,7 @@
 class Admin::RoleAppointmentsControllerTest < ActionController::TestCase
   setup do
-    login_as :writer
+    login_as_preview_design_system_user(:gds_editor)
```
```
  login_as_preview_design_system_user(:writer)
```

Ideally, we should keep the same level of permissions on the user so it's testing the exact same stuff.

davidgisbey commented:
Addressing your questions

```
Is the single drop down option ok for the select? I was unsure how to disable select.
```

If this is a common occurrence, since the component doesn't allow no blank options (i think we should build this in down the line) i think we could do something like:

```
selected:  selected: form.object.person_id == person_id || disambiguated_people_names.count == 1
```
The better option is building a `include_blank` option into the component long term though.

```
Rendering of the 'none' for speeches, and 'Cannot delete' doesn't look great. These are both p tags. Is there an alternative styling I should be using?
```

Carlos settled on this pattern on the people page

<img width="630" alt="image" src="https://user-images.githubusercontent.com/42515961/229715057-83f3e7d0-5825-4071-8e0a-ad6fe79a3d39.png">

I think we should stick with that unless there's a good reason not to.

```
There is a server error if a user tries to submit a blank started at date, but populated ended at date. Presumably to do with model validations, but not sure why the error summary is not handling this?
```

I can se that this is caused by this validator being run and blowing up on nil
https://github.com/alphagov/whitehall/blob/main/app/models/role_appointment.rb#L36-L54

As this is doing a bunch of work to see if there are overlaps it will always blow up if the start date isn't present. There's a bunch of stuff we can do within the validator to check for presence in the conditionals.

However,  it's already going to add an error if start date isn't presence due to this line

```
validates :role_id, :person_id, :started_at, presence: true
```
so the easiest thing to do is probably only run the vallidator if start date is present

```
  validates_with Validator, if: -> {  start_date.present? }
```

chriswhitehouse-tw commented:
cool, thanks for the comments @davidgisbey.

I've implemented text_inset for notes, and i've added the Validator conditional and associated test.

I've not implemented the the suggested code for the drop down, as I don't think it aligns with how the disambiguated_people_names is actually being whittled down to 1. As currently implemented, disambiguated_people_names will always return a full list, it is the select method on line 13 that is responsible to filtering this down to one person if edit page is being rendered, or a list of all people, if the new page is being rendered.

davidgisbey commented:
Thanks for doing that Chris. I'll have a look through this first thing tomorrow am 👍

davidgisbey reviewed (approved):
lgtm!

nnagewad commented:
This looks good. Just wondering about the inset text. Does it appear after the date has been set? Should that be the case? For the interim could the inset text be located above the buttons? Happy to discuss with @chriswhitehouse-tw + @davidgisbey

chriswhitehouse-tw commented:
Thanks for the review @nnagewad. I've updated the code and the screenshots. ✅
